blockstore/apps/bundles/store.py

```diff
@@ -42,6 +43,14 @@ class FileInfo:
     public = attr.ib(type=bool)
     size = attr.ib(type=int)
     hash_digest = attr.ib(type=bytes)
+    mime_type = attr.ib(type=str, default=None)
+
+    @property
+    def data_filename(self):
+        """The data filename with an extension, if present."""
+        hash_hex = self.hash_digest.hex()
+        extension = mimetypes.guess_extension(self.mime_type) if self.mime_type else ''
```

Comment:
I think I see the potential for a rare but possible problem here:

Scenario:
1. User uploads a video file; it's `mime_type` is stored as `video/mp4` and its data file is stored at `[hash].mp4`
2. We update python (or the underlying OS used in the docker image, etc.), and `mimetypes` now returns `.m4v` as the file extension for `video/mp4`
3. Blockstore hasn't stored the file extension, only the mimetype, so now looks for `[hash].mp4` which doesn't exist.


Obviously this isn't super likely, but it has happened before with mimetype projects, e.g. [this change of extension for `audio/mp4` from `.mp4a` to `.m4a`](https://github.com/jshttp/mime-db/commit/ae413e174e9b8d3026b39a43a9a044b72be41ab8) or [this very recent change of the extension for `application/ecmascript` from `.ecma` to `.es`](https://github.com/jshttp/mime-db/commit/03d57a35d71f5897f3a2a9df488c6ef76f34ccc6)

Another related issue is that `mimetypes` can return different results on different systems since it uses data files from the underlying OS.

All that said, it's probably not really a problem in practice and we could always add blockstore-specific overrides for certain mimetypes here if it ever was a problem. Thoughts?

Reply:
@bradenmacdonald:

> I think I see the potential for a rare but possible problem here:
>
> [...]
>
> Blockstore hasn't stored the file extension, only the mimetype, so now looks for [hash].mv4 which doesn't exist.

Oh, that's a nice catch! Loved it. :heart: 

> Thoughts?

This dependency between blockstore behavior and its environment will be painful to maintain. A deployment or upgrade could easily break multiple static file URLs. For example, `XBlockRuntime` relies on [blockstore to translate static URLs](https://github.com/edx/edx-platform/blob/49aaa8621d195a452000942aae8286966a91b96a/openedx/core/djangoapps/xblock/runtime/blockstore_runtime.py#L168) which in turn will use the [snapshot URL generation](https://github.com/edx/blockstore/blob/ece89fa79cd4e4a320f2d8c977032af0a52e2680/blockstore/apps/api/v1/serializers/bundles.py#L152) to retrieve it, and if the extension guessed changes, let's say, because of deployment, so does the static file URL, yielding 400's.

It should be relatively simple to extend the patch to add the "guessed file extension" to the snapshot metadata. But, at this point, I think the patch will be more resilient if we go all the way and save the whole "snapshot filename" instead. Changes to the naming schema won't affect previous Snapshots if we rely on a static filename definition in the snapshot metadata.

Let's get a consensus on this before I submit a new revision, @bradenmacdonald.

Reply:
@jvdm Whoops sorry for the delayed reply here. 

Do you want to merge this now or address this first? I'm fine either way. Although it does seem like a risk of a bug in the future, the odds strike me as very low, and the mitigation would be quite easy (`if mimetype='problematic/type': file_ext = ".forced"; else: file_ext = guess_mime_type()`)

>  I think the patch will be more resilient if we go all the way and save the whole "snapshot filename" instead

Sounds good to me

Reply:
@bradenmacdonald:

> Do you want to merge this now or address this first? I'm fine either way.

Let's merge. We can unblock customers now and follow up on this in a separate PR.

Reply:
OK I'll merge tomorrow so that it deploys during edX business hours.